### ISSUE TITLE
[HOLD for payment 2023-02-03] [$250] Audit forms and fix inconsistencies with focus, tab and shift + tab behavior 2/4

### ISSUE THREAD

**Issue by dylanexpensify** [Expensify/App#7916]
We should audit all our forms and fix any inconsistencies with focus, tab, shift + tab and enter behavior. The expected behavior is as follows:

1. `Tab` navigates to the next input.
2. `Shift + tab` navigates to the previous input.
3. `Enter` submits the form.
4. `Space` toggles checkboxes/dropdowns.

**Note:** We should make sure that tabbing cycles through the form in an order that makes sense, usually top to bottom.

Here's a list of forms to be audited:
- [ ] [RequestorStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/RequestorStep.js) (INTERNAL)
- [ ] [ValidationStep.js](https://github.com/Expensify/App/blob/main/src/pages/ReimbursementAccount/ValidationStep.js) (INTERNAL)
- [ ] [Workspace - Invite new members](https://github.com/Expensify/App/blob/main/src/pages/workspace/WorkspaceInvitePage.js)
- [ ] [Workspace - General settings](https://github.com/Expensify/App/blob/main/src/pages/workspace/WorkspaceSettingsPage.js)
- [ ] [Set password](https://github.com/Expensify/App/blob/main/src/pages/SetPasswordPage.js)

**Upwork job link:** https://www.upwork.com/jobs/~0197285630ce0fa6fa ~https://www.upwork.com/jobs/~01a2053d2c6b3432b4~

**Comment by MelvinBot** [comment]
Triggered auto assignment to @nickmurray47 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @michaelhaxhiu (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by dylanexpensify** [comment]
Note: decision to split this main issue up into smaller issues came from [this convo](https://expensify.slack.com/archives/C01SKUP7QR0/p1645014281039819)

**Comment by MelvinBot** [comment]
@michaelhaxhiu, @nickmurray47 Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by michaelhaxhiu** [comment]
>Note: decision to split this main issue up into smaller issues came from [this convo](https://expensify.slack.com/archives/C01SKUP7QR0/p1645014281039819)

Does this job need to be posted to upwork in a particular fashion basically? Would it be 1 job for everything or 5 jobs (linked to 1 GH issue)?

**Comment by nickmurray47** [comment]
@michaelhaxhiu I personally think it should be 1 job since I assume the fix will be the same for all forms. I can see the argument for splitting it up into multiple jobs but I think whichever contributor suggests the best fix should just do all of the forms in one go.

**Comment by michaelhaxhiu** [comment]
cool thanks nick. @dylanexpensify said the same in a DM.

**Comment by michaelhaxhiu** [comment]
Posted to upwork and added to GH body. Link here too: https://www.upwork.com/jobs/~01a2053d2c6b3432b4

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @thienlnam (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by mdneyazahmad** [comment]
## Proposal
These 4 related issues are very similar. I proposed the solution here https://github.com/Expensify/App/issues/7918#issuecomment-1063298280

**Comment by parasharrajat** [comment]
Reviewing the proposal on another issue.

**Comment by thienlnam** [comment]
Any contributor interested in doing this audit? Involves going through the forms and making sure keyboard actions make sense and making updates to those that don't

**Comment by parasharrajat** [comment]
I am waiting for @mdneyazahmad's new proposal with focus-trap. But I see that this issue is more about audits. @mdneyazahmad Would you like to do the audit?

**Comment by mdneyazahmad** [comment]
@parasharrajat I am sorry I missed the notification and Thank you for the offer. I will soon post the audit of this form. I also think that focus trap should be there. You can see in the video how convenient it is for the user to access modal content with just keyboard. When enter is pressed it should open the modal and focus on the first focusable element. So that he can access the modal immediately on the tab press and when it is closed it should focus it back to where it started. But, this issue is more about audit, we can do it later.



https://user-images.githubusercontent.com/77761491/159399557-31c84de5-1a11-4c3a-a689-476dd9fdc653.mp4

**Comment by parasharrajat** [comment]
Thanks for the suggestion @mdneyazahmad. We will add focus-trap later as a separate issue if needed. This issue only focuses on audits and fixing inconsistencies.

**Comment by michaelhaxhiu** [comment]
@parasharrajat can you confirm if I am correct in understanding that the proposal for this job isn't yet finalized, and thus price should double? 

Or is this a scenario where the proposal is pretty much done and it just needs a little tweak/change?

**Comment by parasharrajat** [comment]
Proposal is pretty much done. Just need to see a new one. 

@mdneyazahmad  I am waiting for your audit proposal.

**Comment by michaelhaxhiu** [comment]
Keeping price at $250 until next week since @mdneyazahmad has an evolving proposal. Let's get that proposal updated/finalized when you have a chance Neyaz

**Comment by mdneyazahmad** [comment]
## Proposal

### RequestorStep
Need help navigating this page.

### ValidationStep
Need help navigating this page.

### Workspace Invitte new members
1. `tab` and `shift + tab` work as expected ✅
2. This form contains `OptionList` accessibility of this component seems broken❌
3. `enter` submits the form ✅

### Workspace - General Settings
1. `tab` and `shift + tab` work as expected ✅
2. When pressing `enter` on Avatar it submits the form ❌
3. `enter` submits the form ✅

### Set password
1. `tab` and `shift + tab` work as expected ✅
4. `checkbox(eye icon)` is toggled with both `enter` and `space` ❌
5.  Set password button is disabled by default until user enters a valid password ❌
6.  Shows error message while typing (once submitted with error) ❌
7. When submitting produces error it does not focus on the element ❌
8. Password input field does not have red outlined for error feedback ❌
9. `enter` submits the form ✅

**Comment by thienlnam** [comment]
Yup, I agree - thanks for the investigation.  Hiring @mdneyazahmad for the job

**Comment by MelvinBot** [comment]
📣 @mdneyazahmad You have been assigned to this job by @thienlnam!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by michaelhaxhiu** [comment]
@mdneyazahmad can you place apply for the job on upwork so I can hire you there?

**Comment by michaelhaxhiu** [comment]
@mdneyazahmad bump on applying for job here - https://www.upwork.com/jobs/~0197285630ce0fa6fa. I know you have several jobs assigned and just wanted to check how it's going?

**Comment by mdneyazahmad** [comment]
I am sorry for being late here. I am working on it and will submit pr by tomorrow. Thank you!

**Comment by mdneyazahmad** [comment]
@michaelhaxhiu applied

**Comment by mdneyazahmad** [comment]
I am struggling to navigate the first two page that seems to be related to bank account flow. Somehow it does not work and gives cors error on `dev`. It works on staging but  it asks me to verify my identity with photo at 3/5 step. Do we have any test credential for `onfido` flow?

I am also facing the same `cors` issue on 1 more related form audit issue. Do need to fix this error on `dev` too or else how are we able to work?

**Comment by michaelhaxhiu** [comment]
good que, let's wait on @thienlnam or @parasharrajat to respond to that one

**Comment by parasharrajat** [comment]
I have the same error on dev.

**Comment by thienlnam** [comment]
@mdneyazahmad Feel free to follow these steps to get access to those forms.   https://expensify.slack.com/archives/C01GTK53T8Q/p1646168719721789?thread_ts=1646163503.484199&cid=C01GTK53T8Q

You'll at least be able to audit those forms on staging but you might have to make changes and then have them tested later on staging

**Comment by mdneyazahmad** [comment]
@thienlnam I am stuck on this step. It asks for me to verify my identity and the two mentioned pages for this issue seems be on step 4 and 5.


https://user-images.githubusercontent.com/77761491/162878789-c67cfa91-00cc-4e76-b00b-02eddd94d4ea.mp4

**Comment by thienlnam** [comment]
@mdneyazahmad Actually, we'll have to leave those bank account forms to an internal audit.  Just address the other forms for now

**Comment by luacmartins** [comment]
@mdneyazahmad please note that `4. Space toggles checkboxes.` was added to the list of expected behavior.

**Comment by mdneyazahmad** [comment]
This issue depends on https://github.com/Expensify/App/issues/7623 I will have to wait until that issue is fixed.

**Comment by parasharrajat** [comment]
@mdneyazahmad Could you please tell us how this depends on that issue?

**Comment by mdneyazahmad** [comment]
In this issue we have to fix inconsistency with `tab` and `shift + tab` which seems to work fine and also the form should submit when enter is pressed.

This depends on that issue because that tries to solve the same core problem how to handle the `Enter` key press and PR for this issue will depend on the decision taken there.

cc: @parasharrajat

**Comment by parasharrajat** [comment]
I think you can move ahead with pressOnEnter in this issue. Do you need anything else?

**Comment by michaelhaxhiu** [comment]
@mdneyazahmad can you provide a quick update here? If you are loaded a little heavy right now, that's ok! but it would be nice to share where things are currently

**Comment by mdneyazahmad** [comment]
@michaelhaxhiu I am ooo and will be back in 1 - 2 days. I am sorry for not informing about it. `tab` and `shift + tab` works as expected. It has some other issue that I mentioned here https://github.com/Expensify/App/issues/7916#issuecomment-1080175468

I think this is the problem here in workspace invite member.
> This form contains OptionList accessibility of this component seems broken❌

1. Arrow key does not work.
2. It does not toggle selection with `space`.
3. It toggles with `enter`.
4. When you press `enter` on selected field it will toggle as well as submit the form (related to the other issue mentioned).


https://user-images.githubusercontent.com/77761491/165665581-ce182bd5-5a89-4581-8c6d-865c1a1804ce.mp4


https://user-images.githubusercontent.com/77761491/165665601-ba9c6881-8f81-4329-b62f-fd749a0b1b70.mp4




https://user-images.githubusercontent.com/77761491/165665825-410478e3-0880-4868-8e37-e026fb3ca9b5.mp4

**Comment by michaelhaxhiu** [comment]
Thanks for the update. Let's wait for @parasharrajat to review those points!

**Comment by parasharrajat** [comment]
I am not very sure of many things so @thienlnam could you please help us finalize those. Let's take each page one by one. 

[Workspace - General settings](https://github.com/Expensify/App/blob/main/src/pages/workspace/WorkspaceSettingsPage.js?rgh-link-date=2022-02-25T14%3A01%3A50Z)

1. Pressing Enter on the avatar opens the Menu as well as submits the form. What is the expected behavior for this case?
2. Space key opens the dropdown when focused. Is this correct?

#### Set Password

1. When the eye icon is focused, pressing Enter toggles it as well well calls the set password action. What is the expected behavior for this case?

### Workspace Invite Page. 

Issues mentioned above. What should be the expected behavior?

**Comment by thienlnam** [comment]
cc @luacmartins Could you help us with the expected behavior of these forms? I would think it's the same for the workspace invite page but wanted to double check

**Comment by luacmartins** [comment]
I think we should standardize on:

- Pressing enter submits the form
- Pressing space interacts with the input, i.e. opens dropdowns, toggles show password, opens the avatar menu, etc.

**Comment by parasharrajat** [comment]
@mdneyazahmad Could you please continue the discussion here to determine the next action?

**Comment by mdneyazahmad** [comment]
@parasharrajat Should we handle the accessibility of `OptionsList` in a separate issue? That is used at many places.

**Comment by parasharrajat** [comment]
I guess so but not my call.

**Comment by mdneyazahmad** [comment]
####  Here's the summary:
- Everything works as expected as per the desired result except (`OptionsList`) https://github.com/Expensify/App/issues/7916#issuecomment-1111680331. I am not sure about the expected result here. That is based on my understanding.
- `Enter` does not work as expected even if you are not focussed on form element and you press `enter` it submits the form sometimes that other element is a general button to go back or close modal (related to https://github.com/Expensify/App/issues/7623).

#### Default accessibility:
1. A button works with click, enter or space.
2. Checkbox is toggled with click or space.
3. Dropdown, we are using a library that seems to work as expected.

As described here https://github.com/Expensify/App/issues/7916#issuecomment-1113247988. I think even if a button(not submit) exists inside the form pressing enter should not submit the form. 

#### Issue with `enter`:
`Button` accepts a prop `pressOnEnter`. When this prop is set, it registers a global listener for key event in capture phase. As soon as this `Button` is mounted it registers a global listener. The result is you are focussed anywhere and press enter it submits the form.

##### Examples:
1. When you are on a form with `<Button pressOnEnter>` and focus on back button press enter goes back as well as submits the form.
2. You are focussed on avatar button and press enter it opens the avatar upload dialog as well as submits the form.

#### Remaining task:
- [ ] Fix form submit on enter.
- [ ] Fix OptionsList component accessibility.
- [ ] Implement custom checkbox for web toggles with click or space. (Current version uses a button that toggles checkbox with click, space or enter).

**Comment by luacmartins** [comment]
Ah I see! Yea, I think in that case we should NOT submit the form and just perform the action of the selected button (on both space and enter press).

**Comment by thienlnam** [comment]
PR in review for fixing checkbox, still awaiting PRs for other items

**Comment by luacmartins** [comment]
@mdneyazahmad is [this issue](https://github.com/Expensify/App/issues/8934) being addressed in one of your PRs?

**Comment by mdneyazahmad** [comment]
Checking the related issues.

**Comment by luacmartins** [comment]
Seems like that issue was fixed already.

**Comment by mdneyazahmad** [comment]
It does not have any issue with (`Tab`  and `Shift + Tab`), PR for checkbox is created here https://github.com/Expensify/App/pull/8874. `Enter` on submit has same issue we are discussing.

cc: @parasharrajat

**Comment by michaelhaxhiu** [comment]
Can we get an update on the latest progress of this issue today, please?

**Comment by michaelhaxhiu** [comment]
I see - we will discuss in slack to arrive at our next steps. Let's post the slack link in that GH after you have posted (just to keep visibility!).

**Comment by mdneyazahmad** [comment]
Asked here https://expensify.slack.com/archives/C01GTK53T8Q/p1654195847140899

**Comment by michaelhaxhiu** [comment]
@mdneyazahmad do you mind responding to @roryabraham on the Slack Discussion that you posted/linked above?

**Comment by michaelhaxhiu** [comment]
Slack discussion is still hatching.

**Comment by michaelhaxhiu** [comment]
held up still, seems we need a verdict and path forward from the slack discussion.

**Comment by roryabraham** [comment]
Posted a possible solution on the slack thread and in https://github.com/Expensify/App/pull/9749

**Comment by michaelhaxhiu** [comment]
Still in progress!

**Comment by michaelhaxhiu** [comment]
Not overdue fam, but let's bump slack and/or the PR https://github.com/Expensify/App/pull/9749

**Comment by michaelhaxhiu** [comment]
We are still discussing in slack https://expensify.slack.com/archives/C01GTK53T8Q/p1659459279504269?thread_ts=1654195847.140899&cid=C01GTK53T8Q

**Comment by thienlnam** [comment]
Same as the other audit forms issues - we should have reached a conclusion on that slack, could we also get an update here @mdneyazahmad

**Comment by mdneyazahmad** [comment]
Consensus in the thread is to move forward with:

> "Continue using a single global keyboard event handler library, patching it with the solution from https://github.com/Expensify/App/pull/9749 so that local event handlers can override the global ones"

Working on [this PR](https://github.com/Expensify/App/pull/9749)

**Comment by thienlnam** [comment]
Any updates on this @mdneyazahmad ?

**Comment by michaelhaxhiu** [comment]
Not overdue, chill out melvin-bot

**Comment by michaelhaxhiu** [comment]
PR still open, and waiting for review by rajat

**Comment by michaelhaxhiu** [comment]
>Hoping for an update in 2 days on this one, as the testing list is big.

**Comment by michaelhaxhiu** [comment]
Rajat is testing https://github.com/Expensify/App/pull/10965#issuecomment-1274839311

**Comment by parasharrajat** [comment]
PR is under review.

**Comment by michaelhaxhiu** [comment]
Seems PR is still pending. Are we close @parasharrajat?

**Comment by parasharrajat** [comment]
Awaiting update on the PR.

**Comment by arielgreen** [comment]
@michaelhaxhiu just updated the label on this to daily! What needs to happen to get this over the finish line?

**Comment by parasharrajat** [comment]
Update: Old PR had some issues which we ended up closing. A new PR is created which needs to be reviewed.

**Comment by michaelhaxhiu** [comment]
Noticed that PR was closed yesterday. This seems to be taking quite a while to complete. Is that mainly because this is harder, or what's slowing it down? 

Commented on the PR too asking for a rough ETA of completion @mdneyazahmad 🙏

**Comment by michaelhaxhiu** [comment]
Ok got it. @parasharrajat Are you able to do that?

**Comment by parasharrajat** [comment]
Yup, I will take a look after https://github.com/Expensify/App/pull/12546.

**Comment by kevinksullivan** [comment]
@parasharrajat do you have an ETA for reviewing this issue?

**Comment by michaelhaxhiu** [comment]
Rajat mentioned to me in a DM:
>⁦I will get it done this weekend. I hope Author will be active for feedback.

**Comment by michaelhaxhiu** [comment]
Re-applying the `Bug` label to re-assign, as I'm headed OOO starting this weekend.

**Comment by parasharrajat** [comment]
PR is under active review now. Will take to end line asap :muscle:

**Comment by michaelhaxhiu** [comment]
Woo nice!

**Comment by parasharrajat** [comment]
Suggested PR improvements and requested more details on the changes. I will approve as soon as my doubts are cleared.

**Comment by puneetlath** [comment]
I'm just catching up here. Are all of these changes being handled in one PR?

**Comment by mdneyazahmad** [comment]
@puneetlath thanks for catching this up. There is one PR deployed which is common. Here's the final PR (https://github.com/Expensify/App/pull/12370). You can read the context section on the PR description about that. Thanks!

**Comment by puneetlath** [comment]
Thanks @mdneyazahmad. @parasharrajat looks like this is now waiting on you, is that right?

**Comment by puneetlath** [comment]
@mdneyazahmad @parasharrajat can you guys provide a summary of where we're at with this? And also a link to the PR?

**Comment by JmillsExpensify** [comment]
Also I'm left wondering whether we should instead close this issue? It's focused on tabbing behavior and other keyboard navigation that is not a focus of our bug fixing machine.

**Comment by JmillsExpensify** [comment]
At the very least, it's not a focus for WAQ so putting a hold on this issue until we agree on next steps.

**Comment by mdneyazahmad** [comment]
PR https://github.com/Expensify/App/pull/12370 is slowly moving on. It was stuck on loading page.

**Comment by puneetlath** [comment]
Do we think we can get it to the finish line this week?

**Comment by puneetlath** [comment]
@parasharrajat @luacmartins @thienlnam how close do you think we are on this one? Can we get it closed off in the next couple of days?

**Comment by puneetlath** [comment]
My sense is that there wasn't really a PR that caused a specific bug. Rather we developed form guidelines and then went and implemented them. Do you agree with that @parasharrajat @thienlnam?

**Comment by puneetlath** [comment]
For me: 
@mdneyazahmad contract: https://www.upwork.com/nx/wm/workroom/30032384/overview
@parasharrajat contract: https://www.upwork.com/nx/wm/pre-hire/c/8577561/offer/23159468

**Comment by puneetlath** [comment]
All paid. Thanks for your contributions everyone!
